### DISCUSSION TITLE
Discussion: Type inference workflow can be confusing for static member in generic type

### DISCUSSION THREAD

**Discussion by smoothdeveloper** [dotnet/fsharp#6605] [General]
#### Repro steps

I have hit this one several times:

```fsharp
type Rec<'T> = { a : 'T }
with
    static member Empty (* : Rec<'T> *) = { a = Unchecked.defaultof<_> }

let getA (a:Rec<int>) = a.a
//let getEmptyA = getA Rec.Empty
//let e = Rec<exn>.Empty
```

Initially forgetting type annotation on the static member.

I think this is painful for newcomers and a reasonably annoying issue for more seasoned users.

I understand the reason the tooling type inference would pick up `int` in this case, despite the compiled type of `Rec<'T>.Empty` is always going to be `Rec<obj>`.

My main issue is that, at no point in my workflow, have I seen a warning on the `Empty` member declaration.

#### Actual behavior

![confusing warnings typeinference in tooling versus compile](https://user-images.githubusercontent.com/87944/56491725-57abd480-649e-11e9-8b45-82baa2c59e7e.gif)

#### Expected behavior

I'd like to figure out a way to get a warning to show up there as early as possible, maybe a check on specific or more generic cases can be added that would spawn such warning.

One issue is that it would be a breaking change when code uses implicit/inferred fallback to `obj` and maybe other cases where type inference will pick a single instantiation.

Ideas for checks:

specific one:
* In generic type
* static member with return type of declaring type
* no explicit type annotation

generic one:
* member returning generic type instantiation whose type arguments are not specified or can't be derived by any input parameter

would those checks be sound? other ideas?

Note:
* focusing on static member first, I believe this happens more frequently, looking for feedback
* may be relevant to FsLint, but I'm wondering how users faced several times beside having good understanding.

#### Known workarounds

Hit the issue frequently enough in a row to not hit it for a while

#### Related information 

Win10/VS2019

**Comment by cartermp** [comment]
This is really a consequence of the pit of failure you can get into with wildcard generics. Note that explicitly using the type parameter in the `Empty` member makes the compiler happy in all cases.

I'm not sure if this is a beginner problem; wildcard generic type arguments are kind of an advanced feature that I doubt too many people really use. I personally try to avoid them since I don't like trying to predict if I'm going to end up with a value restriction somewhere.

**Comment by smoothdeveloper** [comment]
Thanks for the feedback @cartermp, it makes sense that using `Unchecked.defaultof<_>` is the initial issue, but I've hit the same without using any explicit wildcard generics:

https://github.com/Microsoft/visualfsharp/blob/3533a60a25f26f31d79c1634a7702a83a9f697c9/src/fsharp/MethodCalls.fs#L102-L111

I was missing type annotation on `Empty` member, and no wildcard written, does that make sense?

This may be very limitied use case not worth considering in context of compiler, but really looking to see if other people are also hiting this and if there are ideas how to make the compiler more helpful troubleshooting this.
